### ISSUE TITLE
[Bug]:  Error receiving complete

### ISSUE THREAD

**Issue by andriux1990** [opensourcepos/opensourcepos#4138]
### Bug Description?

What does this line of code do: $supplier = $supplier->get_info($supplier_id); //TODO: supplier is never used after this. in the model, within the public function save_value? It seems that this line causes issues when completing a sale.

I want to work on the receivings module, add AJAX to avoid the annoying page reloads, and also include the option to change the selling price at the time of purchase. Additionally, I aim to update the interface to improve the user experience.

If I make these changes, would you accept them to be included in the project?

### Steps to Reproduce?

.

### Expected Behavior?

.

### OpensourcePOS Version

development (unreleased)

### Php version

Php 8.3

### Server Operating System and version

Linux

### Database Management System and version

Mariadb

### Web Server and version

Apache

### System Information Report (optional)

System Information Report

### Unmodified copy of OpensourcePOS

- [X] I agree this copy has not been modified

**Comment by andriux1990** [comment]
In the ospos_receivings table, we can add a field called receiving_status to manage the status of a purchase. This would allow suspending a purchase and completing it later when needed.

ALTER TABLE `ospos_receivings`
  ADD COLUMN IF NOT EXISTS `receiving_status` tinyint(2) NOT NULL DEFAULT 0;

**Comment by jekkos** [comment]
well I believe we should then see how sale suspending works and try to unify the code as much as possible. Sounds great to improve things a bit, but we"ll need to go one thing at a time.
So it's probably best to work in a focused manner and go step by step, like one feature per PR. Otherwise the scope is too big and it will be hard to review.

UI improvements like AJAX calls could be a good start.

**Comment by andriux1990** [comment]
I already have the AJAX functionality for switching modes between entry, return, and requisition. I implemented it for requisition or transfers, but I found the problem that it's not possible to add items when it's in requisition mode.

But this problem is in the code of ospos. I will review the issue and try to fix it.

**Comment by andriux1990** [comment]
In requisition the problem is that the values ​​passed to form_input in the view are not always strings, as functions like to_currency_no_money or to_quantity_decimals can return floating point numbers.

view / receivings
<tbody id="cart_contents">
			<?php
			if(count($cart) == 0)
			{
			?>
				<tr>
					<td colspan='9'>
						<div class='alert alert-dismissible alert-info'><?= lang('Sales.no_items_in_cart') ?></div>
					</td>
				</tr>
			<?php
			}
			else
			{
				foreach(array_reverse($cart, true) as $line => $item)
				{
			?>
					<?= form_open("$controller_name/editItem/$line", ['class' => 'form-horizontal', 'id' => "cart_$line"]) ?>
						<tr>
							<td><?= anchor("$controller_name/deleteItem/$line", '<span class="glyphicon glyphicon-trash"></span>') ?></td>
							<td><?= esc($item['item_number']) ?></td>
							<td style="text-align:center;">
								<?= esc($item['name'] . ' ' . implode(' ', [$item['attribute_values'], $item['attribute_dtvalues']])) ?><br />
								<?= '[' . to_quantity_decimals($item['in_stock']) . ' in ' . $item['stock_name'] . ']' ?>
								<?= form_hidden('location', (string)$item['item_location']) ?>
							</td>

							<?php if ($items_module_allowed && $mode != 'requisition') { ?>
								<td><?= form_input([
									'name' => 'price',
									'class' => 'form-control input-sm',
									'value' => (string)to_currency_no_money($item['price']),
									'onClick' => 'this.select();'
								]) ?></td>
							<?php } else { ?>
								<td>
									<?= (string)$item['price'] ?>
									<?= form_hidden('price', (string)to_currency_no_money($item['price'])) ?>
								</td>
							<?php } ?>

							<td><?= form_input([
								'name' => 'quantity',
								'class' => 'form-control input-sm',
								'value' => (string)to_quantity_decimals($item['quantity']),
								'onClick' => 'this.select();'
							]) ?></td>
							<td><?= form_dropdown(
								'receiving_quantity',
								$item['receiving_quantity_choices'],
								(string)$item['receiving_quantity'],
								['class' => 'form-control input-sm']
							) ?></td>

							<?php if ($items_module_allowed && $mode != 'requisition') { ?>
								<td>
									<div class="input-group">
										<?= form_input([
											'name' => 'discount',
											'class' => 'form-control input-sm',
											'value' => (string)($item['discount_type'] ? to_currency_no_money($item['discount']) : to_decimals($item['discount'])),
											'onClick' => 'this.select();'
										]) ?>
										<span class="input-group-btn">
											<?= form_checkbox([
												'id' => 'discount_toggle',
												'name' => 'discount_toggle',
												'value' => 1,
												'data-toggle' => "toggle",
												'data-size' => 'small',
												'data-onstyle' => 'success',
												'data-on' => '<b>' . $config['currency_symbol'] . '</b>',
												'data-off' => '<b>%</b>',
												'data-line' => $line,
												'checked' => $item['discount_type'] == 1
											]) ?>
										</span>
									</div>
								</td>
							<?php } else { ?>
								<td><?= (string)$item['discount'] ?></td>
								<?= form_hidden('discount', (string)$item['discount']) ?>
							<?php } ?>

							<td>
								<?= to_currency(
									($item['discount_type'] == PERCENT)
										? $item['price'] * $item['quantity'] * $item['receiving_quantity'] - $item['price'] * $item['quantity'] * $item['receiving_quantity'] * $item['discount'] / 100
										: $item['price'] * $item['quantity'] * $item['receiving_quantity'] - $item['discount']
								) ?>
							</td>
							<td><a href="javascript:$('#<?= esc("cart_$line", 'js') ?>').submit();" title=<?= lang(ucfirst($controller_name) . '.update') ?>><span class="glyphicon glyphicon-refresh"></span></a></td>
						</tr>
						<tr>
							<?php
							if($item['allow_alt_description'] == 1)	//TODO: ===?
							{
							?>
								<td style="color: #2F4F4F;"><?= lang('Sales.description_abbrv').':' ?></td>
							<?php
							}
							?>
							<td colspan='2' style="text-align: left;">
								<?php
								if($item['allow_alt_description'] == 1)	//TODO: ===?
								{
									echo form_input ([
										'name' => 'description',
										'class' => 'form-control input-sm',
										'value' => $item['description']
									]);
								}
								else
								{
									if ($item['description'] != '')	//TODO: !==?
									{
										echo $item['description'];
										echo form_hidden('description', $item['description']);
									}
									else
									{
										echo '<i>'.lang('Sales.no_description').'</i>';
										echo form_hidden('description','');
									}
								}
								?>
							</td>
							<td colspan='7'></td>
						</tr>
					<?= form_close() ?>
			<?php
				}
			}
			?>
		</tbody>

**Comment by andriux1990** [comment]
I resolved the issues related to the requisition functionality, focusing on the view, controller, and model.
I adapted the function in the controller and added a new one in the model.
The transfer was successfully completed, updating the origin and destination locations by deducting and adding the values accordingly. Additionally, in ospos_inventory, I added the corresponding records for REQ IN and REQ OUT.

![Screenshot 2024-12-27 at 00-07-39 localhost _ localhost _ mobile _ ospos_inventory phpMyAdmin 5 2 1](https://github.com/user-attachments/assets/e7b28461-695d-40fa-976e-0ce73868df54)
